### PR TITLE
[2370] Hook up back links for new course flow

### PR DESCRIPTION
**Pull request by danielburnley**
preview

### Context

When going through the new course flow, the user should be able to see and use a `back` link to go to their previous steps, with the following circumstances:

- Going through the flow normally
- Going back to the confirmation page
- Going back past locations if there is only a single location
- Going back to different pages depending on the user being an accredited body

### Changes proposed in this pull request

- Add in correct back links
- Add in view helper for doing back links with confirmation

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

**Comment by danielburnley** [comment]
https://trello.com/c/M6J2ViCg/2370-s-hook-up-back-links-for-new-course-flow

**Comment by misaka** [comment]
Feels like a good candidate to have a quick gif demonstrating how this works. @TobyRet what were you using to do that, again?

**Comment by danielburnley** [comment]
now with added gif @misaka (OBS -> ffmpeg to gif is a way to do it on linux, apparently)

File: app/controllers/courses/accredited_body_controller.rb [hunk 1]
```diff
@@ -99,6 +108,14 @@ def build_course_params
       @autocompleted_provider_code = params[:course].delete(:autocompleted_provider_code)
     end
 
+    def build_back_link
+      @back_link_path = if @provider.sites.count > 1
+                          new_provider_recruitment_cycle_courses_locations_path(course: @course_creation_params)
+                        else
+                          new_provider_recruitment_cycle_courses_study_mode_path(course: @course_creation_params)
+                        end
+    end
```
Comment (misaka) [inline app/controllers/courses/accredited_body_controller.rb]: This `build_back_link`, and all the other versions of it in the other controllers, is doing my head in. That is, until I have a look at `NextCourseCreationStepService`.

I feel like we should have a single service that handles both `next` and `back` ... I know that breaks the principle of having an `execute` as a single point of entry for services ... maybe we could have `Providers::CourseCreation::NextStepService` and `Providers::CourseCreation::BackService`, with the commonalities kept in `Providers::CourseCreation`? That might be messy if we have to define `NextStepService` and `BackService` in a separate Ruby file, which would be necessary for the auto-loader to work (and Rubocop not to complain, probably).

Not sure what's the best approach, but I do think that this needs a bit of cleanup, it's not clear what the logic is doing here.

Comment (defong) [reply app/controllers/courses/accredited_body_controller.rb]: Agreed, its doing my head as well.

``` ruby
def oscillate(current, course) do
{
  previous: previous_link_factory,
  next: next_link_factory,
  roulette_spin: lucky_day 
}
end

```

**Comment by misaka** [comment]
> now with added gif @misaka (OBS -> ffmpeg to gif is a way to do it on linux, apparently)

How much work is it?

File: app/controllers/courses/accredited_body_controller.rb (hunk 1 above)
Comment (danielburnley) [reply app/controllers/courses/accredited_body_controller.rb]: Yeah I'll make a service first, then I'll look at some common interface for getting both the next and back links :thinking: I did avoid making a service for this reason, it's a bit of a headache to get a ncie way to do it but i'll dedicate some time to it

**Comment by faissalMT** [comment]
OBS is straightforward enough, ffmpeg might be a bit fiddly though https://engineering.giphy.com/how-to-make-gifs-with-ffmpeg/

**Comment by misaka** [comment]
Ran into a small problem: accredited bodies aren't re-selected when I go back to them from a later page.

Steps:

1. navigate to provider 2HN's course page
2. create a new course (by append /new to the URL)
3. when on the accredited bodies select an existing provider as the accredited provider
4. click "Continue"
5. use the "Back" link to go back to the previous page
6. :boom: the previously-selected accredited body is no longer selected

Notably, when using the browser back-button this persists.

Unless this is trivial to fix, we should probably just deploy this ticket as-is and create a new bug ticket for this.

**Pull request merged**